src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -0,0 +1,221 @@
+                }
+            }
+
+            public ILocalRepositoryModel CreateRepository(string path) => new LocalRepositoryModel(path);
+        }
+
+        void Refresh(object gitExt)
+        {
+            try
+            {
+                var repo = FindActiveRepository(gitExt);
+                var newSolutionPath = dte?.Solution?.FullName;
+
+                if (repo == null && newSolutionPath == solutionPath)
+                {
+                    // Ignore when ActiveRepositories is empty and solution hasn't changed.
+                    // https://github.com/github/VisualStudio/issues/1421
+                    log.Information("Ignoring no ActiveRepository when solution hasn't changed");
+                }
+                else
+                {
+                    var newRepositoryPath = repo?.RepositoryPath;
+                    var newBranchName = repo?.BranchName;
+                    var newHeadSha = repo?.HeadSha;
+                    var newTrackedSha = newRepositoryPath != null ? service.FindTrackedSha(newRepositoryPath) : null;
+
+                    if (newRepositoryPath != repositoryPath)
+                    {
+                        log.Information("Fire PropertyChanged event for ActiveRepository");
+                        ActiveRepository = repo != null ? service.CreateRepository(repo.RepositoryPath) : null;
+                        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ActiveRepository)));
+                    }
+                    else if (newBranchName != branchName)
+                    {
+                        log.Information("Fire StatusChanged event when BranchName changes for ActiveRepository");
+                        StatusChanged?.Invoke(this, EventArgs.Empty);
+                    }
+                    else if (newHeadSha != headSha)
+                    {
+                        log.Information("Fire StatusChanged event when HeadSha changes for ActiveRepository");
+                        StatusChanged?.Invoke(this, EventArgs.Empty);
+                    }
+                    else if (newTrackedSha != trackedSha)
+                    {
+                        log.Information("Fire StatusChanged event when TrackedSha changes for ActiveRepository");
+                        StatusChanged?.Invoke(this, EventArgs.Empty);
+                    }
+
+                    repositoryPath = newRepositoryPath;
+                    branchName = newBranchName;
+                    headSha = newHeadSha;
+                    solutionPath = newSolutionPath;
+                    trackedSha = newTrackedSha;
+                }
+            }
+            catch (Exception e)
+            {
+                log.Error(e, "Refreshing active repository");
+            }
+        }
+
+        static RepositoryInfo FindActiveRepository(object gitExt)
+        {
+            var activeRepositoriesProperty = gitExt.GetType().GetProperty("ActiveRepositories");
```

Comment:
Instead of passing `gitExt` as an object and doing all this manual work here, how about wrapping it in an interface of our own and exposing a type that this code can safely use?

Reply:
There is already a `IVSGitExt` (https://github.com/github/VisualStudio/blob/master/src/GitHub.Exports/Services/IVSGitExt.cs https://github.com/github/VisualStudio/blob/master/src/GitHub.TeamFoundation.14/Services/VSGitExt.cs) type that represents (and wraps) `IGitExt`, so you can use that, there's no need for this code. If you prefer to take it out of the TeamFoundation assemblies and use reflection in it, you should be able to do that easily in it as well.

Reply:
The fact that `TeamExplorerContext` uses `IGitExt` is a detail that I didn't want to expose in a public API. `ITeamExplorerContext` is intended to surface the pattern of events that we need and hide the gory details.

I'm hoping that `IVSGitExt` and the parts of `ITeamExplorerServiceHolder` that use it will no longer be required.

I'm about to push a version that wraps `gitExt` in a private class and encapsulates the reflection code. I want to avoid exposing this as a public API.

Reply:
Actually something that just came to me: could/should we use `dynamic` here instead of reflection?